### ISSUE TITLE
Perl does not tend to gobble up all the arguments that follow it

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14959]
Migrated from [rt.perl.org#126250](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=126250) (status was 'rejected')

Searchable as RT126250$

**Comment by p5pRT** [comment]
### From @KES777

#### Created by @KES777



Acording to \[doc\]\(http&#8203;://perldoc\.perl\.org/perlop\.html\#Terms\-and\-List\-Operators\-%28Leftward%29\)
\*\*list operators tend to gobble up all arguments that follow\*\*

\#\!/usr/bin/env perl

use warnings;

\#sub test;
\#sub My&#8203;::test;

sub list \{
&nbsp;   print ">>@&#8203;\_\<\<\\n";
\};

my $obj =  bless \{\}\, 'main';
list\( 0\, test $obj\, 1\, 2\, 3 \);
list\( 0\, My&#8203;::test $obj\, 1\, 2\, 3 \);

\# sub test \(@&#8203;\) \{ \# it has not matter there is prototype or not
sub test \{
&nbsp;   print "test>>@&#8203;\_\<\<\\n";

&nbsp;   return;
\}


new\(\);


package My;

\# sub test \(@&#8203;\) \{ \# it has not matter there is prototype or not
sub test \{
&nbsp;   print "My&#8203;::test>>@&#8203;\_\<\<\\n";
&nbsp;   return;
\}


In this example the output is&#8203;:
test>>main=HASH\(0x24a3160\)\<\<
>>0 1 2 3\<\<
My&#8203;::test>>main=HASH\(0x24a3160\)\<\<
>>0 1 2 3\<\<

Which means that list does not gobble up all the arguments that follow it
The expecting result is&#8203;:
test>>main=HASH\(0x1bea160\) 1 2 3\<\<
>>0\<\<
My&#8203;::test>>main=HASH\(0x1bea160\) 1 2 3\<\<
>>0\<\<

Digging deeper I realise that here comes into play the indirect object call\.
http&#8203;://perldoc\.perl\.org/perlobj\.html\#Invoking\-Class\-Methods

list 0\, test main 1\, 2\, 3;
list 0\, My&#8203;::test $obj 1\, 2\, 3;

which produce
test>>main 1 2 3\<\<
>>0\<\<
My&#8203;::test>>main=HASH\(0x159e160\) 1 2 3\<\<
>>0\<\<

But here I do not want to use indirect syntax call and do not expect that behavior here\.

Give precendence and let to gobble up all the arguments

Or\, at least\, if the behavior will not be fixed\,
we must warn here how it was resolved\. Like&#8203;:

&nbsp;    Resolved as indirect method call&#8203;: $obj\->test\(\)\.
&nbsp;    Use paranthes if you mean test\( $obj\, 1\, 2\, 3 \)


NOTICE&#8203;:
when I do test call in void context I do not get all the warnings&#8203;:

my $obj =  bless \{\}\, 'main';
test $obj\, 1\, 2\, 3;

Useless use of a constant \(2\) in void context at \./t2\.pl line 15\.
Useless use of a constant \(3\) in void context at \./t2\.pl line 15\.
test>>main=HASH\(0x1714160\)\<\<

The '1' is also in void context\, so here I also expect :
Useless use of a constant \(1\) in void context at \./t2\.pl line 15\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.22.0:

Configured by kes at Tue Sep 29 16:12:55 EEST 2015.

Summary of my perl5 (revision 5 version 22 subversion 0) configuration:
   
  Platform:
    osname=linux, osvers=3.13.0-37-generic, archname=x86_64-linux
    uname='linux keswork 3.13.0-37-generic #64-ubuntu smp mon sep 22 21:28:38 utc 2014 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Dprefix=/home/kes/perl'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.8.4', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/4.8/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.19.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.19'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'



@INC for perl 5.22.0:
    /home/kes/perl5/lib/perl5/x86_64-linux
    /home/kes/perl5/lib/perl5
    /home/kes/perl/lib/site_perl/5.22.0/x86_64-linux
    /home/kes/perl/lib/site_perl/5.22.0
    /home/kes/perl/lib/5.22.0/x86_64-linux
    /home/kes/perl/lib/5.22.0
    .


Environment for perl 5.22.0:
    HOME=/home/kes
    LANG=ru_UA.UTF-8
    LANGUAGE=en
    LC_MESSAGES=en_US.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/kes/perl/bin:/home/kes/perl/bin:/home/kes/perl5/bin:/home/kes/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games
    PERL5LIB=/home/kes/perl5/lib/perl5
    PERL_BADLANG (unset)
    PERL_MB_OPT=--install_base "/home/kes/perl5"
    PERL_MM_OPT=INSTALL_BASE=/home/kes/perl5
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cowens

On Fri\, Oct 2\, 2015 at 6&#8203;:25 AM KES \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> Acording to \[doc\]\(http&#8203;://perldoc\.perl\.org/perlop\.html\#Terms\-and\-List\-Operators\-%28Leftward%29\)
> \*\*list operators tend to gobble up all arguments that follow\*\*

> \# sub test \(@&#8203;\) \{ \# it has not matter there is prototype or not

This is because prototypes only count if you declare them before their
use\.  This is also why things are working the way you expect\.  You are
leaving perl to guess what the barewords test and My&#8203;::test mean\.  If
you had declared them as subs you would have gotten the result you
desired\.  Instead\, perl made the best guess it could\.

> Which means that list does not gobble up all the arguments that follow it
> The expecting result is&#8203;:
> test>>main=HASH\(0x1bea160\) 1 2 3\<\<
> >>0\<\<
> My&#8203;::test>>main=HASH\(0x1bea160\) 1 2 3\<\<
> >>0\<\<
>

That isn't the expected result even if we were to ignore indirect
object syntax\.  The test function hasn't been declared\, so what you
said \(ignoring indirect object syntax\) was equivalent to this&#8203;:

list 0\, "test" $obj\, 1\, 2\, 3

which produces

Scalar found where operator expected at q\.pl line 9\, near ""test" $obj"
&nbsp;        \(Missing operator before  $obj?\)
Number found where operator expected at q\.pl line 9\, near "$obj 1"
&nbsp;        \(Missing operator before 1?\)



> But here I do not want to use indirect syntax call and do not expect that behavior here\.

Then don't let perl guess at things\.  Always either declare or define
your functions before you use them\.

>
> Give precendence and let to gobble up all the arguments
>

Definitely the wrong choice as it would require parsing the whole file
to find the functions\, then a second parse to actually build the
opcodes\.

> Or\, at least\, if the behavior will not be fixed\,
> we must warn here how it was resolved\. Like&#8203;:
>
>     Resolved as indirect method call&#8203;: $obj\->test\(\)\.
>     Use paranthes if you mean test\( $obj\, 1\, 2\, 3 \)

This might be a good idea if it could be implemented in a way that
does not slow down parsing significantly\.

>
> NOTICE&#8203;:
> when I do test call in void context I do not get all the warnings&#8203;:
>
> my $obj =  bless \{\}\, 'main';
> test $obj\, 1\, 2\, 3;
>
> Useless use of a constant \(2\) in void context at \./t2\.pl line 15\.
> Useless use of a constant \(3\) in void context at \./t2\.pl line 15\.
> test>>main=HASH\(0x1714160\)\<\<
>
> The '1' is also in void context\, so here I also expect :
> Useless use of a constant \(1\) in void context at \./t2\.pl line 15\.

I \_think\_ this is side effect of not warning for 1; in some contexts&#8203;:

perl \-we '1\, 2\, 3'
Useless use of a constant \(2\) in void context at \-e line 1\.
Useless use of a constant \(3\) in void context at \-e line 1\.

perl \-we '2\, 3\, 4'
Useless use of a constant \(2\) in void context at \-e line 1\.
Useless use of a constant \(3\) in void context at \-e line 1\.
Useless use of a constant \(4\) in void context at \-e line 1\.

This lack of a warning is useful because of code like

eval \{
&nbsp;    do\_something\(\);
&nbsp;    1;
\} or die "it blew up&#8203;: $\_";

and the end of modules \(which are required to have a true value as the
last statement\)\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @mauke

Am 02\.10\.2015 um 15&#8203;:58 schrieb Chas\. Owens&#8203;:
>
> I \_think\_ this is side effect of not warning for 1; in some contexts&#8203;:
>
> perl \-we '1\, 2\, 3'
> Useless use of a constant \(2\) in void context at \-e line 1\.
> Useless use of a constant \(3\) in void context at \-e line 1\.
>
> perl \-we '2\, 3\, 4'
> Useless use of a constant \(2\) in void context at \-e line 1\.
> Useless use of a constant \(3\) in void context at \-e line 1\.
> Useless use of a constant \(4\) in void context at \-e line 1\.
>
> This lack of a warning is useful because of code like
>
> eval \{
>      do\_something\(\);
>      1;
> \} or die "it blew up&#8203;: $\_";
>
> and the end of modules \(which are required to have a true value as the
> last statement\)\.

Those aren't in void context\, though\. And '0;' is also exempt from warnings\.

It is useful in cases like '1 while do\_stuff\(\)'\, which has 1 in void 
context\. \(This is an idiom but personally I think '\(\) while \.\.\.' might 
have been more obvious as a dummy value \(at the cost of 1 extra 
character :\-\)\.\)


\-\- 
Lukas Mai \<plokinom@&#8203;gmail\.com>

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Oct 02\, 2015 at 09&#8203;:58&#8203;:27AM \-0400\, Chas\. Owens wrote&#8203;:
> On Fri\, Oct 2\, 2015 at 6&#8203;:25 AM KES \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > Or\, at least\, if the behavior will not be fixed\,
> > we must warn here how it was resolved\. Like&#8203;:
> >
> >     Resolved as indirect method call&#8203;: $obj\->test\(\)\.
> >     Use paranthes if you mean test\( $obj\, 1\, 2\, 3 \)
> 
> This might be a good idea if it could be implemented in a way that
> does not slow down parsing significantly\.

I think there are enough legitimate uses of such syntax that such warnings
would be more annoying than useful\. For example consider code that checked
the return value of print&#8203;:

&nbsp;    sub ok \{ die unless $\_\[0\] \}

&nbsp;    ok print $fh 1\,2\,3;

\-\- 
Never do today what you can put off till tomorrow\.

**Comment by p5pRT** [comment]
### From @KES777

ok\, I understand about warnigs now\. Never meet that in docs yet \(\(

05\.10\.2015\, 11&#8203;:10\, "Dave Mitchell via RT" \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> On Fri\, Oct 02\, 2015 at 09&#8203;:58&#8203;:27AM \-0400\, Chas\. Owens wrote&#8203;:
>>  On Fri\, Oct 2\, 2015 at 6&#8203;:25 AM KES \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>  > Or\, at least\, if the behavior will not be fixed\,
>>  > we must warn here how it was resolved\. Like&#8203;:
>>  >
>>  > Resolved as indirect method call&#8203;: $obj\->test\(\)\.
>>  > Use paranthes if you mean test\( $obj\, 1\, 2\, 3 \)
>>
>>  This might be a good idea if it could be implemented in a way that
>>  does not slow down parsing significantly\.
>
> I think there are enough legitimate uses of such syntax that such warnings
> would be more annoying than useful\. For example consider code that checked
> the return value of print&#8203;:
>
>     sub ok \{ die unless $\_\[0\] \}
>
>     ok print $fh 1\,2\,3;
>
> \-\-
> Never do today what you can put off till tomorrow\.

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'rejected'
